### PR TITLE
etcd3: use PrevKV to remove additional get

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: #https://github.com/kubernetes/kubernetes/issues/33653

We are trying to test using PrevKV feature and see if it improves performance.
In order to test this, we will need etcd v3.1 (alpha) image.

Blockers:
- update gcr.io image (version v3.0.12)

``` release-note
etcd3: use PrevKV to remove additional get
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34246)

### CODE REVIEW

**Comment by hongchaodeng** [comment]
cc @wojtek-t @timothysc

**Comment by xiang90** [comment]
Do not merge this. If this solves the problem, we will back port this feature to 3.0 release branch next week.

File: pkg/storage/etcd3/watcher.go [hunk 1]
```diff
@@ -154,7 +154,15 @@ func (wc *watchChan) sync() error {
...
 
 	for _, kv := range getResp.Kvs {
-		wc.sendEvent(parseKV(kv))
+		prevResp, err := wc.watcher.client.Get(wc.ctx, string(kv.Key), clientv3.WithRev(kv.ModRevision-1))
+		if err != nil {
+			return err
+		}
+		var prevVal []byte
+		if len(prevResp.Kvs) > 0 {
+			prevVal = prevResp.Kvs[0].Value
```
Comment (hongchaodeng) [inline pkg/storage/etcd3/watcher.go]: "0" will returns latest state.

**Comment by wojtek-t** [comment]
@xiang90 @hongchaodeng - just to clarify, what I should do is to:
- build etcd Docker image for 3.1 release
- patch this PR
- build from this patch and check the results
Or am I missing something?

**Comment by hongchaodeng** [comment]
Yes. Plan matched.

**Comment by xiang90** [comment]
> build from this patch and check the results

If the result is good, we will backport this feature back to etcd 3.0.

**Comment by wojtek-t** [comment]
@hongchaodeng @xiang90 - I did an experiment with this PR patched (and etcd in 3.1.0-alpha.1 version - I build the container: "gcr.io/google_containers/etcd:3.1.0-alpha.1" for this purpose)

And the results are waaaaaaaay better - please take a look into this comment:
https://github.com/kubernetes/kubernetes/issues/33653#issuecomment-252218603

So if you could backport this feature to 3.0.x next week it would be great !!!

**Comment by wojtek-t** [comment]
I'm reassigning it back to you. Once you build the new 3.0.x release, we should probably update this PR and then merge it, right?

**Comment by hongchaodeng** [comment]
@wojtek-t 
Sure. I will update the PR once the new release with the backport comes.

**Comment by timothysc** [comment]
@wojtek-t will need to update gcr.io container to support this change.

**Comment by hongchaodeng** [comment]
@timothysc @wojtek-t 
Verified on v3.0.12.
We need to update to that.

**Comment by wojtek-t** [comment]
@timothysc @hongchaodeng - we need to build an image to make it possible to run real clusters. But this PR doesn't change anything for real clusters - in particular godep verification is failing. So there are still some issues with this PR (though e.g. kubemark failure seems to be unrelated).

**Comment by hongchaodeng** [comment]
@wojtek-t 
I will fix the godep license failure.
Can you also try this patch and verify about the perf results with etcd 3.0.12?

**Comment by hongchaodeng** [comment]
@timothysc @wojtek-t 
The check failure is "godep verify" problem.
See logs:
```
godep: Dep (github.com/coreos/etcd/vendor/github.com/coreos/pkg/capnslog) restored, but was unable to load it with error:
	Package (github.com/coreos/etcd/vendor/github.com/coreos/pkg/capnslog) not found
```
It assumes "github.com/coreos/etcd/vendor/github.com/coreos/pkg/capnslog" is a package. But it's not. It belongs to vendor.
etcd 3.0.12 didn't use any vendor tool to save vendor.
etcd 3.1 use glide.

**Comment by wojtek-t** [comment]
@thockin @lavalamp - any thoughts about the above? I'm not expert in godeps, and we really need this PR.

**Comment by wojtek-t** [comment]
Also @hongchaodeng - this was exactly the same case for 3.0.10 etcd release - you didn't change anything in the meantime wrt to vendoring. So we are doing something wrong in this PR.

@timothysc - you did the previous upgrade to 3.0.10. Maybe you can describe how exactly you did:
https://github.com/kubernetes/kubernetes/pull/33393, so that we can do the same here?

**Comment by wojtek-t** [comment]
@thockin @lavalamp - never mind, I figured it out.

@hongchaodeng - please take a look into #34440 (which is passing tests and is built on top of your PR). What you need to do here is to:
- remove all the "new" dependencies in Godeps.json (just leave the lines that were modified)
- remove all changes to "licenses" file
- also please squash all the godeps-related commits

[I would already merge #34440, but it seems there are some cla-related issues if I'm picking commits from someone outside Google.]

Anyway - if you apply the changes I described above, I will lgtm this PR.
@xiang90 - FYI

**Comment by wojtek-t** [comment]
BTW - for the purpose of tests, I just pushed the "3.0.12-experimental.1" image to gcr (I didn't want to push the official one yet, since we may be doing some changes connected to migration).

**Comment by wojtek-t** [comment]
@hongchaodeng - you also need to remove the "licenses" that you added

**Comment by hongchaodeng** [comment]
@wojtek-t 
Thanks. I run update-godep-license.sh again.

**Comment by wojtek-t** [comment]
@hongchaodeng - but the linceses are still there, maybe you didn't push the PR?

**Comment by hongchaodeng** [comment]
Just "git push"...

**Comment by wojtek-t** [comment]
lgtm - thanks

**Comment by hongchaodeng** [comment]
@k8s-bot gce e2e test this

**Comment by hongchaodeng** [comment]
@k8s-bot gce e2e test this

**Comment by hongchaodeng** [comment]
@k8s-bot gce e2e test this

**Comment by hongchaodeng** [comment]
Running into this issue many times:
```
error running up: exit status 1
```

**Comment by hongchaodeng** [comment]
@k8s-bot gce e2e test this

**Comment by hongchaodeng** [comment]
@k8s-bot gce e2e test this

**Comment by thockin** [comment]
regarding godep:  did you follow the EXACT instructions in https://github.com/kubernetes/kubernetes/blob/master/docs/devel/godep.md ?

If so, please help me with instructions for the smallest repro I can make...

**Comment by hongchaodeng** [comment]
@thockin 
Thanks for offering the help.
We have fixed it actually :)

**Comment by wojtek-t** [comment]
@k8s-bot gce e2e test this, issue: #IGNORE (Failed to create firewall)

**Comment by timothysc** [comment]
@smarterclayton - this patch is high-priority.

**Pull request merged**